contracts/LimitsChecker.sol

```diff
@@ -0,0 +1,267 @@
+    /// @param _limit Limit to set
+    /// @param _periodDurationMonth  Length of period in months to set.
+    /// Duration of the period can be 1, 2, 3, 6 or 12 months.
+    function setLimitParameters(uint256 _limit, uint256 _periodDurationMonth)
+        external
+        onlyRole(SET_LIMIT_PARAMETERS_ROLE)
+    {
+        _checkPeriodDurationMonth(_periodDurationMonth);
+        periodDurationMonth = uint64(_periodDurationMonth);
+        currentPeriodEnd = uint128(_getPeriodEndFromTimestamp(block.timestamp));
+        limit = uint128(_limit);
+        if (spent > limit) {
+            spent = limit;
+        }
+
+        emit LimitsParametersChanged(_limit, _periodDurationMonth);
+    }
+
+    /// @notice Returns limit and periodDurationMonth
+    /// @return limit - the maximum that can be spent in a period
+    /// @return periodDurationMonth - length of period in months
+    function getLimitParameters() external view returns (uint256, uint256) {
+        return (limit, periodDurationMonth);
+    }
+
+    /// @notice Returns amount spent in the current period, balance available for spending,
+    /// @notice start date of the current period and end date of the current period
+    /// @return _alreadySpentAmount - amount spent in the current period
+    /// @return _spendableAmount - balance available for spending in the current period
+    /// @return _periodStartTimestamp - start date of the current period
+    /// @return _periodEndTimestamp - end date of the current period
+    function getCurrentPeriodState()
+        external
+        view
+        returns (
+            uint256 _alreadySpentAmount,
+            uint256 _spendableAmount,
+            uint256 _periodStartTimestamp,
+            uint256 _periodEndTimestamp
+        )
+    {
+        return (
+            spent,
+            limit - spent,
+            _getPeriodStartFromTimestamp(currentPeriodEnd - 1),
+            currentPeriodEnd
+        );
+    }
+
+    // ------------------
+    // PRIVATE METHODS
+    // ------------------
+
+    /// currentPeriodEnd is calculated as a calendar date of the beginning of a next month, bi-months, quarter, half year, or year period.
+    /// If, for example, periodDurationMonth = 3, then it is considered that the date changes once a quarter.
+    /// And can take values 01.01, 01.04, 01.07, 01.10.
+    /// Then at the moment when it is necessary to shift the currentPeriodEnd (the condition is fulfilled: block.timestamp >= currentPeriodEnd),
+    /// currentPeriodEnd takes on a new value and spent is set to zero. Thus begins a new period.
+    function _checkAndUpdateLimitParameters() internal {
+        _checkPeriodDurationMonth(periodDurationMonth);
+        if (block.timestamp >= currentPeriodEnd) {
+            currentPeriodEnd = uint128(_getPeriodEndFromTimestamp(block.timestamp));
+            spent = 0;
+        }
+    }
+
+    function _checkPeriodDurationMonth(uint256 _periodDurationMonth) internal view {
```

Comment:
I guess that if we allow having four months-wide period also then it could be easily simplified to:
```
require((_periodDurationMonths > 0) && (_periodDurationMonths < 12) && (_periodDurationMonths % 12 == 0), ...)
```

Reply:
I suppose it might be a business-level requirement, though, yes, this is a bit vague, indeed.

The second minor point is that some pieces of code rely on the fact that the period always fits into the same year (easy to fix).